server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -709,12 +709,14 @@ public final DocumentParserContext createNestedContext(NestedObjectMapper nested
      * Return a new context that has the provided document as the current document.
      */
     public final DocumentParserContext switchDoc(final LuceneDocument document) {
-        return new Wrapper(this.parent, this) {
+        DocumentParserContext cloned = new Wrapper(this.parent, this) {
             @Override
             public LuceneDocument doc() {
                 return document;
             }
         };
+        cloned.inArrayScope = false;  // Reset for new document.
```

Comment:
Maybe add the reason why to reset array scope for nested docs?

Reply:
Updated. 

The second pass doesn't work for nested objects in general, as it adds ignored source to the wrong doc.. It should be extremely rare to need it, it'd be an array within an array within a nested object, and a leaf tracked in ignored source.. 

Updating nested docs in the second pass would be very tricky. We should revisit this if we start tracking parsing scopes.